src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -249,16 +246,30 @@ public void leaveToken(DetailAST ast) {
     /**
      * Checks whether a class should be declared as final or not.
      *
-     * @param desc description of the class
+     * @param classDesc description of the class
      * @return true if given class should be declared as final otherwise false
      */
-    private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
-        return desc.isWithPrivateCtor()
-                && !(desc.isDeclaredAsAbstract()
-                    || desc.isSuperClassOfAnonymousInnerClass())
-                && !desc.isDeclaredAsFinal()
-                && !desc.isWithNonPrivateCtor()
-                && !desc.isWithNestedSubclass();
+    private static boolean shouldBeDeclaredAsFinal(ClassDesc classDesc) {
+        final boolean isAbstract = classDesc.isDeclaredAsAbstract();
+        final boolean isSuperClassOfAnonymousInnerClass =
+                classDesc.isSuperClassOfAnonymousInnerClass();
+        final boolean declaredAsFinal = classDesc.isDeclaredAsFinal();
+        final boolean withNestedSubClass = classDesc.isWithNestedSubclass();
+        final boolean shouldBeFinal;
+
+        if (classDesc.isImplicitConstructor()) {
```

Comment:
Code should be read as English text.

`if  classDesc is Implicit Constructor` is nonsense.

Try:
`if  classDesc has Implicit Constructor`

Reply:
@romani I have changed with `isWithImplicitConstructor` according to the ci failure in previous push only hasImplicitConstructor will not look good as ishasImplicitConstructor